### ISSUE TITLE
NLL: complete transition from migrate to full NLL

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#57895]
This is a tracking issue for the work items remaining for the NLL team before we can consider NLL stabilized across all editions (see also #43234).

Here is an outline of the steps we have identified.

 * [x] Stage 1: Move all editions to NLL (#57804)
    * Move Rust 2015 to borrowck=migrate
    * Potential blockers:
      * NLL-sound bugs that affect migration mode (vs ones that only arise with `#![feature(nll)]` today)
      * Major diagnostic regressions that affect migration mode
 * [ ] Stage 2: Transition fully to NLL (#58781)
   * one currently opts into this via `#![feature(nll)]` or `-Z borrowck=mir -Z two-phase-borrows`)
   * this causes NLL warnings (downgraded from errors) become errors again
     * possible intermediate steps
       * move to deny-by-default lints for the warnings 
       * move 2018 to hard error on its own, leaving 2015 with downgrade-to-warnings
   * It also gets rid of AST-region inference, causing some code to be accepted that is rejected by migrate mode.
   * Potential blockers: 
     * NLL-sound bugs that affect NLL mode without migration
     * Major diagnostic regressions
     * Two-phase borrows (2PB) question or other "unsettled questions" must have some conservative answer in place
     * Crater run showing no major crates fail in practice
 * [x] Stage 3: remove old AST-borrowck (done in https://github.com/rust-lang/rust/pull/64790)

### Blocking issues

* [ ] https://github.com/rust-lang/rust/issues/73154
* [x] #57642 
* [x] #59159

**Comment by pnkfelix** [comment]
triage: P-high. Assigning to self.

**Comment by pnkfelix** [comment]
While discussing #59114 at the T-lang design team meeting, one question that arose was whether we would move to hard error mode for both 2015 and 2018 at the same time, or if we would do it in a staged fashion (presumably on 2018 first and then following with 2015 sometime later).

Discussion of that pointed out the detail that there are some languge features that depend, in some fashion, on NLL.

An incomplete list follows:
 * bind_by_move_pattern_guards #15287 
 * if-let guards
 * certain kinds of array length const evaluations? (need to double-check with @centril) 

The question then became whether those language feature depend solely on NLL migrate mode, or if they depend on NLL hard error mode. In particular, some features are deliberately choosing not to implement AST-borrowck support; in such cases, I can imagine an error under NLL migrate mode leading to an ICE (or worse, unsoundly accepting the code using a new feature with a mere warning) arising from migrate mode's attempt to run AST-borrowck when there was no support for that feature implemented in AST-borrowck.

This provides some support for the argument that we should move to full NLL mode in a staged fashion, doing only 2018 edition first (and the 2018 can add support for the aforementioned language features), and waiting to move to full NLL on the 2015 edition later.

**Comment by Centril** [comment]
(updated ^ with links)

**Comment by pnkfelix** [comment]
I no longer think this is a P-high issue. It *is* something we want to do, and something I plan to keep track of, but it is not something that warrants visiting every week at the compiler meeting.

triage: P-medium.

**Comment by Centril** [comment]
- NLL Error => Warning downgrades are removed in https://github.com/rust-lang/rust/pull/64221.
- AST borrowck is removed in https://github.com/rust-lang/rust/pull/64790.

**Comment by Centril** [comment]
I believe the only things that remain now are dealing with https://github.com/rust-lang/rust/issues/57642 and https://github.com/rust-lang/rust/issues/59159 and then we can remove migrate mode.

**Comment by jackh726** [comment]
Closing in favor of #58781 and #43234
